jupyter-extension/jupyterlab_pachyderm/pps_client.py

```diff
@@ -41,72 +39,142 @@ async def create(self, path: str, config: dict):
-        if 'input' not in config['run']:
-            raise HTTPError(reason=f"Bad Request: field run.input not set")
-
-        input_spec = config['run'].pop("input")  # Input not allowed in SAME metadata.
-        # Paths are relative to the config file. Since the config is being written into
-        #  a temporary file, we need to specify an absolute path.
-        config['notebook']['path'] = os.path.join(os.getcwd(), os.path.relpath(path, '/'))
-        config['notebook']['name'] = os.path.basename(path)
-
-        if 'requirements' in config['notebook'] and not config['notebook']['requirements']:
-            # Remove requirements field if none are specified.
-            del config['notebook']['requirements']
-
-        with NamedTemporaryFile() as temp_config:
-            Path(temp_config.name).write_text(json.dumps(config))
-            command = [sys.executable, "-m", "sameproject.main", "run"]
-            command.extend(["--same-file", temp_config.name])
-            command.extend(["--target", "pachyderm"])
-            command.extend(["--input", input_spec])
-            call = subprocess.run(command, capture_output=True)
-            if call.returncode:
-                raise HTTPError(
-                    code=500,
-                    log_message=call.stderr.decode(),
-                    reason="Internal Error: SAME could not create pipeline"
-                )
+        config = PpsConfig.from_json(config)
+        path = path.lstrip("/")
+        if not os.path.exists(path):
+            raise HTTPError(reason=f"notebook does not exist: {path}")
+        if config.requirements and not os.path.exists(config.requirements):
+            raise HTTPError(reason="requirements file does not exist")
+
+        script, _resources = self.nbconvert.from_filename(path)
+
+        client = python_pachyderm.Client()  # TODO: Auth?
+        companion_repo = f"{config.pipeline_name}__context"
+        if companion_repo not in (item.repo.name for item in client.list_repo()):
+            client.create_repo(
+                repo_name=companion_repo,
+                description=f"files for running the {config.pipeline_name} pipeline"
+            )
+
+        companion_branch = upload_environment(client, companion_repo, config, script.encode('utf-8'))
+        pipeline_spec = create_pipeline_spec(config, companion_branch)
+        client.create_pipeline_from_request(
+            python_pachyderm.parse_dict_pipeline_spec(pipeline_spec)
+        )
 
         return json.dumps(
             dict(message=None)  # We can send back console link here.
         )
 
 
-def create_pipeline_spec(
-    pipeline_name: str,
-    image: str,
-    input_spec: dict,
-    script_name: str,
-) -> dict:
-    """Generates the pipelines spec. Currently copied from """
-    companion_repo = f"{pipeline_name}__context"
-    micro_entrypoint = (
-        'print("Greetings from the Pachyderm PPS Extension"); '
-        + 'import sys; '
```

Comment:
It looks like we've changed how we're running the script? (This wrapper doesn't look present in the new version)

Reply:
It has changed. This method that was deleted basically defines our entrypoint in as a series of python statements and is hacky, not very maintainable, and would most likely be confusing for the user when looking at their pipeline spec. However, when moving that code over from SAME, I realized we should just upload that entrypoint as a script within the companion repo and actually doing so removes a lot of the complication surrounding it.